remove replace directives

carnott-snap opened the pull request:
go.mod contains many replace entries that all consumers must include to
import logic. These are cancerous and tend to make consumers builds
difficult or outright fail. All have been removed, except for the one
required by flux. This entry, while retained, has converted into a noop
and should be removed when fluxcd/flux#2590 is fixed.

carnott-snap committed:
remove replace directives

go.mod contains many replace entries that all consumers must include to
import logic. These are cancerous and tend to make consumers builds
difficult or outright fail. All unnecessary replaces have been removed.

One is required by flux and should be removed when
docker/distribution#2905 is released.

The second is caused by a co-dependant transitive upgrade that bumps
k8s.io/code-generator beyond kubernetes-1.14.4. To correct this the
go.mod has been left untidy, with the correct version required and
replaced so that build/test will pass.

carnott-snap commented on go.mod:
```diff
@@ -1,47 +1,42 @@
+//             +-> k8s.io/code-generator@v0.0.0-20190511023357-639c964206c2
+//
+// This replace should be removed once fluxcd/flux is released.
+replace k8s.io/code-generator => k8s.io/code-generator v0.0.0-20190311093542-50b561225d70
```
removal of this dependency cycle will require synchronised commits in both `helm-operator` and `flux`:
1. remove all `require` and `replace` blocks from helm-operator: carnott-snap/helm-operator@100d4a6e2252dc3a988c1a3aec2dcbdccc560763
2. pin `flux` to pseudo version of `helm-operator`: carnott-snap/flux@65da8ef3744eaa2b70445f2fcaaee95b8d033ea6
3. remove unnecessary `replace` blocks:
    - carnott-snap/helm-operator@d8142acbe99b9d93a4a4ea8e5ff1e06118427903
    - carnott-snap/flux@a8f13dff40162434e67c324eebbb6905033c52d7

carnott-snap commented:
> Thanks, but this belongs to a separate PR.

When you say this, what would you like to see teased apart?

I think you are referring to the migration from `github.com/weaveworks/flux/git` to `github.com/fluxcd/flux/pkg/git`. We can certainly do that first, but I think it is important to make that a prerequisite to the removal of the `replace` directives, since that commit is the root of the kubernetes issues.

> Also, I am not sure we should rewrite the history of the Changelogs, at the time of the release those tickets/PRs lived in the weaveworks repo (and the links are not broken, since GitHub redirects automatically).

My `find -exec sed` may have been more aggressive than you would like, but I think that it serves to stabilise the links, since at a later date `weaveworks` could create a `flux` repo and break that redirect. Github does not guarantee the redirect after a migration, just offers it when it is possible.

2opremio commented:
Yes, I am referring to the renames. They don't belong to this PR. I am still unsure of whether we should merge them but, if anything, they should be placed in a separate PR.

carnott-snap commented:
> Yes, I am referring to the renames. They don't belong to this PR. I am still unsure of whether we should merge them but, if anything, they should be placed in a separate PR.

I have isolated and reduced that change in #101, and will wait till that merges before we proceed.

Are you interested in restarting the dependency cycle as described above, so that we can cleanup the `require`s? Based on the work in fluxcd/flux#2590, there should be a way to maintain the `go.sum` without the entries.

If we cannot come to an understanding on removing them, I would be interested to fixup the `go.mod`, and dependency graph, so that no `require`s are needed for customers.

hiddeco commented:
@carnott-snap as you may have noticed I am refactoring large parts (if not all) of the code this touches in #99. Would you be open to make this PR target the `helm-v3-dev` branch to make the future merge of `master` with the ongoing development branch for support of Helm v3 easier?

It is not likely that there will be a release anyway before those two have been merged together, so release wise it does not make a difference.

carnott-snap commented:
> @carnott-snap as you may have noticed I am refactoring large parts (if not all) of the code this touches in #99. Would you be open to make this PR target the `helm-v3-dev` branch to make the future merge of `master` with the ongoing development branch for support of Helm v3 easier?

I am happy to make the change against `helm-v3-dev`, however the inclusion of `k8s.io/kubernetes@v1.15.4` is going to complicate things substantially. As such, I will make a separate pr to track that branch, and we should be able to land both.

carnott-snap commented:
@hiddeco, after further review, it appears that `helm-operator` will be unable to consume the `helm.sh/helm/v3` module beyond `v3.0.0-beta.3`. I think we should push forward with this change and review `helm-v3-dev` when it has been updated to upstream@latest.

This issue is caused by your `replace` block downgrading `k8s.io/kubectl` and `k8s.io/apiextensions-apiserver` to a version below what `helm.sh/helm/v3` needs:
```
github.com/fluxcd/helm-operator/pkg/helm/v3 imports
        helm.sh/helm/v3/pkg/chartutil imports
        k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1: module k8s.io/apiextensions-apiserver@latest found (v0.0.0-20191109110701-3fdecfd8e730, replaced by k8s.io/apiextensions-apiserver@v0.0.0-20190918201827-3de75813f604), but does not contain package k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1
github.com/fluxcd/helm-operator/pkg/helm/v3 imports
        helm.sh/helm/v3/pkg/kube imports
        k8s.io/kubectl/pkg/cmd/util: module k8s.io/kubectl@latest found (v0.0.0-20191111223430-169c48cabbe1, replaced by k8s.io/kubectl@v0.0.0-20190602132728-7075c07e78bf), but does not contain package k8s.io/kubectl/pkg/cmd/util
github.com/fluxcd/helm-operator/pkg/helm/v3 imports
        helm.sh/helm/v3/pkg/kube imports
        k8s.io/kubectl/pkg/validation: module k8s.io/kubectl@latest found (v0.0.0-20191111223430-169c48cabbe1, replaced by k8s.io/kubectl@v0.0.0-20190602132728-7075c07e78bf), but does not contain package k8s.io/kubectl/pkg/validation
github.com/fluxcd/helm-operator/pkg/helm/v3 imports
        helm.sh/helm/v3/pkg/kube tested by
        helm.sh/helm/v3/pkg/kube.test imports
        k8s.io/kubectl/pkg/cmd/testing: module k8s.io/kubectl@latest found (v0.0.0-20191111223430-169c48cabbe1, replaced by k8s.io/kubectl@v0.0.0-20190602132728-7075c07e78bf), but does not contain package k8s.io/kubectl/pkg/cmd/testing
```

hiddeco commented:
@carnott-snap can you revise this PR so that it is mergeable with master, and write down the steps that you have in mind that should happen after this has been merged? As it is hard for me to determine what has and has not been done at this point. Thanks :1st_place_medal:
